nixos/hardware/printers: stop cupsd when unneeded

rnhmjoj opened the pull request:
###### Description of changes

If socket activation is enabled (the default) and printers are configured
declaratively, the ensure-printers service will always start cupsd and
leave it running, thus defeating the point of socket activation.

With this change ensure-printers continues to start the cups.service at
boot, but automatically stops it afterwards if socket activation is
enabled.

Note: Later restarts of ensure-printers will also restart cupsd, but
it's not an issue since it will be reactivate, if necessary.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- Tested `hardware.printers.ensurePrinters`
  - [x] with socket activation
  - [x] without socket activation
- [ ] Tested compilation of all packages that depend on this change (mass rebuild!)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
cups: fix cups.socket unit

The cups.socket unit shouldn't be part of cups.service: stopping the
service would stop the socket and break subsequent socket activations.

See https://github.com/apple/cups/issues/6005

rnhmjoj committed:
nixos/hardware/printers: stop cupsd when unneeded

If socket activation is enable (the default) and printers are configured
declaratively, the ensure-printers service will always start cupsd and
leave it running, thus defeating the point of socket activation.

With this change ensure-printers continues to start the cups.service at
boot, but automatically stops it afterwards if socket activation is
enabled.

Note: Later restarts of ensure-printers will also restart cupsd, but
it's not an issue since it will be reactivate, if necessary.

rnhmjoj committed:
nixos/cupsd: stop managing /run/cups directory

The directory is already set up by cups.socket: managing it with
RuntimeDirectory in cups.service is unnecesary and has the unwanted
effect that upon stopping cupsd systemd will remove it. This includes
the /run/cups/cups.sock socket, so it breaks socket activation.

RaitoBezarius commented:
Looks good: do we have NixOS test for cupsd? It would be nice to introduce one to test this kind of things.

rnhmjoj commented:
> Looks good: do we have NixOS test for cupsd? It would be nice to introduce one to test this kind of things.

There is (un)fortunately. I've been banging my head against it trying to figure out why it breaks.
For some very mysterious reason the socket-activated server (but not the client) loses its printers after restarting the service.

rnhmjoj committed:
nixos/cupsd: fix /var/run/ warning in cups.socket

Silence the systemd warning

    ListenStream= references a path below legacy directory /var/run/, updating /var/run/cups/cups.sock → /run/cups/cups.sock; please update the unit file accordingly

by resetting the value coming from the upstream unit.

rnhmjoj committed:
nixos/tests/printing: split into service/socket

This splits the tests into two: one where cups.socket is started
normally, the order with socket activation.

Why? It's almost impossible to follow the test with 4 different
machines printing at the same time. It should also be more efficient
because only two VMs at a time were needed anyway.

rnhmjoj commented:
Ok, I figured it out: there's a dumb "stateless" mode that just wipes all cups directories on starting the service. I tried a bunch of hacks, including wrapping cupsd and hijacking `sd_notify` communication, to try to set up the printers during socket activation, but it's not possible. With the stateless mode the service must be left running, as it currently does.
